Fix for issue #5086

bummzack opened the pull request:
Improve `Member_Validator` to:
- properly check for existing members.
- allow extensions.
- remove old code and replace with new syntax and add config API.

Fix issue in Group code where Member_Validator was instantiated via "new" which didn't allow injector overrides.
Added unit-tests.

bummzack commented:
@dhensby Ok, this one should be good if CI passes.

dhensby commented on security/Member.php:
```diff
@@ -1842,31 +1860,35 @@ public function __construct() {
+			}
+		}
+
+		$members = Member::get()->filter($identifierField, $data[$identifierField]);
```
What if `$data[$identifierField]` is not set?

bummzack replied:
You mean, if somebody removes the identifier field from the list of required fields? Seems unlikely, but better be safe than sorry… should I create a new PR?

dhensby replied:
No. You can just amend this PR and force push. The updates will come into the PR automatically

dhensby commented:
see https://github.com/silverstripe/silverstripe-framework/pull/5088

bummzack commented:
Is there something I missed from the discussion in #5088?

dhensby commented:
No, in just cross linking so that people can see the discuss ons

bummzack commented:
@dhensby Thanks. I forgot about amending commits. I updated the commit with a safeguard check for `$data[$identifierField]`. I also went ahead and implemented the behavior I proposed in #5086 to solve cases where the validator might not have an `ID` passed with `$data`, yet still "belongs" to a Member.

bummzack commented:
Any idea what this test failure means: 'Failed asserting that 'JoeEdited' is not equal to <string:JoeEdited>.' ?

dhensby commented:
I've restarted those tests, bizarre that a couple have it and a couple don't not really sure.

The behat test is more worrying. If you check this out locally and test, does it work in the CMS when you add/save/edit/delete members?

bummzack commented:
Yeah, everything works fine in the CMS. The warning dialogue now shows up underneath the E-Mail field though… maybe that's causing a problem there?

dhensby commented:
hmm - I don't think that'd be a problem

bummzack commented:
Strangely enough all tests have passed with my previous PR… not sure why it failed now. When testing in the CMS, everything works/validates as expected.

dhensby commented:
someone with more knowledge of PHPUnit will have to explain this one :/

bummzack commented:
@dhensby I figured out the problem. It is important that the validator checks the `ID` passed via `$data` first. Only if there's no ID passed via `$data` or via controller->record, the member assigned to the Validator should come into play.

dhensby commented:
Thanks - I'm pretty happy with this and it could lay the foundation for skipping validation on default admins too

dhensby commented:
@tractorcow thoughts?

dhensby commented:
oh - as it's a new feature, maybe it'll have to go against 3 :/

bummzack commented:
You mean, it's a feature that the `Member_Validator` is suddenly working? For me this is more of a Bugfix :)

tractorcow commented:
Yeah sorry, I meant to say this on the other thread: Can you raise on the 3 branch since it introduces new method API?

tractorcow commented:
Code is pristine by the way.

bummzack commented:
Ok, PR against 3 branch is here: #5098. Sorry for the many PRs
